Sources/NIOWebSocket/NIOWebSocketFrameAggregator.swift

```diff
@@ -0,0 +1,138 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+import NIO
+
+
+/// `NIOWebSocketFrameAggregator` buffers inbound fragmented `WebSocketFrame`'s and aggregates them into a single `WebSocketFrame`.
+/// It guarantees that a `WebSocketFrame` with an `opcode` of `.continuation` is never forwarded.
+/// Frames which are not fragmented are just forwarded without any processing.
+/// Fragmented frames are unmasked, concatenated and forwarded as a new `WebSocketFrame` which is either a `.binary` or `.text` frame.
+/// `extensionData`, `rsv1`, `rsv2` and `rsv3` are lost if a frame is fragmented because they cannot be concatenated.
+/// - Note: `.ping`, `.pong`, `.closeConnection`frames are forwarded during frame aggregation
+public final class NIOWebSocketFrameAggregator: ChannelInboundHandler {
+    enum Error: Swift.Error {
+        case nonFinalFragmentSizeIsTooSmall
+        case tooManyFragments
+        case accumulatedFrameSizeIsTooLarge
+        case receivedNewFrameWithoutFinishingPrevious
+        case didReceiveFragmentBeforeReceivingTextOrBinaryFrame
+    }
+    public typealias InboundIn = WebSocketFrame
+    public typealias InboundOut = WebSocketFrame
+    
+    private let minNonFinalFragmentSize: Int
+    private let maxAccumulatedFrameCount: Int
+    private let maxAccumulatedFrameSize: Int
+    
+    private var bufferedFrames: [WebSocketFrame] = []
+    private var accumulatedFrameSize: Int {
+        bufferedFrames
+            .map({ $0.length })
+            .reduce(0, +)
+    }
+    
+    
+    /// Configures a `NIOWebSocketFrameAggregator`.
+    /// - Parameters:
+    ///   - minNonFinalFragmentSize: Minimum size in bytes of a fragment which is not the last fragment of a complete frame. Used to defend agains many really small payloads.
+    ///   - maxAccumulatedFrameCount: Maximum number of fragments which are allowed to result in a complete frame.
+    ///   - maxAccumulatedFrameSize: Maximum accumulated size in bytes of buffered fragments. It is essentially the maximum allowed size of an incoming frame after all fragments are concatenated.
+    public init(
+        minNonFinalFragmentSize: Int,
+        maxAccumulatedFrameCount: Int,
+        maxAccumulatedFrameSize: Int
+    ) {
+        self.minNonFinalFragmentSize = minNonFinalFragmentSize
+        self.maxAccumulatedFrameCount = maxAccumulatedFrameCount
+        self.maxAccumulatedFrameSize = maxAccumulatedFrameSize
+    }
+    
+    
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        let frame = unwrapInboundIn(data)
+        do {
+            switch frame.opcode {
+            case .continuation:
+                guard let firstFrameOpcode = bufferedFrames.first?.opcode else {
+                    throw Error.didReceiveFragmentBeforeReceivingTextOrBinaryFrame
+                }
+                try bufferFrame(frame)
```

Comment:
A mild nit here and throughout this function: NIO prefers to use the `self` keyword explicitly where possible.

Reply:
Done. Do you have a documented code style guide?

Reply:
I'm afraid we don't. Our style is fairly lenient most of the time. One day Swift will decide on a common style and then we'll take a big jump over to that but until then we're kinda stuck doing this.